src/gui/EditWidgetIcons.cpp

```diff
@@ -158,107 +153,79 @@ void EditWidgetIcons::setUrl(const QString& url)
-                    ++m_redirectCount;
-                    fetchFavicon(m_redirectUrl);
-                } else {
-                    // website is trying to redirect to itself or
-                    // maximum number of redirects has been reached, fall back to Google
-                    fetchFaviconFromGoogle(url.host());
-                }
-            } else {
-                fetchFaviconFromGoogle(url.host());
-            }
-        });
-    });
-
-    if (!requestMade) {
-        resetFaviconDownload();
-        return;
-    }
-
-    m_httpClient->setConnectingTimeOut(5000, [this]() {
-        QUrl tempurl = QUrl(m_url);
-        if (tempurl.scheme() == "http") {
-            resetFaviconDownload();
-            emit messageEditEntry(tr("Unable to fetch favicon.") + "\n" +
-                                  tr("Hint: You can enable Google as a fallback under Tools>Settings>Security"),
-                                  MessageWidget::Error);
+    // Attempt to simply load the favicon.ico file
+    QImage image = fetchFavicon(url);
+    if (!image.isNull()) {
+        addCustomIcon(image);
+    } else if (config()->get("security/IconDownloadFallbackToGoogle", false).toBool()) {
+        QUrl faviconUrl = QUrl("https://www.google.com/s2/favicons");
+        faviconUrl.setQuery("domain=" + QUrl::toPercentEncoding(url.host()));
+        // Attempt to load favicon from Google
+        image = fetchFavicon(faviconUrl);
+        if (!image.isNull()) {
+            addCustomIcon(image);
         } else {
-            tempurl.setScheme("http");
-            m_url = tempurl.url();
-            tempurl.setPath("/favicon.ico");
-            fetchFavicon(tempurl);
+            emit messageEditEntry(tr("Unable to fetch favicon."), MessageWidget::Error);
         }
-    });
+    } else {
+        emit messageEditEntry(tr("Unable to fetch favicon.") + "\n" +
+                              tr("Hint: You can enable Google as a fallback under Tools>Settings>Security"),
+                              MessageWidget::Error);
+    }
 
-    m_ui->faviconButton->setDisabled(true);
+    m_ui->faviconButton->setDisabled(false);
+#endif
 }
 
-void EditWidgetIcons::fetchFaviconFromGoogle(const QString& domain)
+#ifdef WITH_XC_NETWORKING
+size_t writeCurlResponse(char* ptr, size_t size, size_t nmemb, void* data)
 {
-    if (config()->get("security/IconDownloadFallbackToGoogle", false).toBool() && m_fallbackToGoogle) {
-        resetFaviconDownload();
-        m_fallbackToGoogle = false;
-        QUrl faviconUrl = QUrl("https://www.google.com/s2/favicons");
-        faviconUrl.setQuery("domain=" + QUrl::toPercentEncoding(domain));
-        fetchFavicon(faviconUrl);
-    } else {
-        resetFaviconDownload();
-        emit messageEditEntry(tr("Unable to fetch favicon."), MessageWidget::Error);
-    }
+    QByteArray* response = static_cast<QByteArray*>(data);
+    size_t realsize = size * nmemb;
+    response->append(ptr, realsize);
```

Comment:
Since we know the real size, you can use response->reserve() to pre-allocate memory.

Reply:
Actually reserve doesn't make sense in this context because we are adding the new data received to an existing buffer. The append function is accurate.

Reply:
True. I somehow added a loop in my head, which isn't actually there.